phd-tests/tests/src/boot_order/efi_utils.rs

```diff
@@ -0,0 +1,490 @@
+            (2, 1) => {
+                // ACPI Device Path
+                let size = bytes.read_u16::<LittleEndian>()?;
+                if size != 0xc {
+                    bail!("ACPI Device Path size is wrong (was {:#04x}, not 0x000c)", size);
+                }
+                let hid = bytes.read_u32::<LittleEndian>().unwrap();
+                let uid = bytes.read_u32::<LittleEndian>().unwrap();
+                Ok(DevicePath::Acpi { hid, uid })
+            }
+            (1, 1) => {
+                // PCI device path
+                let size = bytes.read_u16::<LittleEndian>()?;
+                if size != 0x6 {
+                    bail!("PCI Device Path size is wrong (was {:#04x}, not 0x0006)", size);
+                }
+                let function = bytes.read_u8().unwrap();
+                let device = bytes.read_u8().unwrap();
+                Ok(DevicePath::Pci { device, function })
+            }
+            (4, 6) => {
+                // "PIWG Firmware File" aka "Firmware File" in UEFI PI spec
+                let size = bytes.read_u16::<LittleEndian>()?;
+                if size != 0x14 {
+                    bail!(
+                        "Firmware File size is wrong (was {:#04x}, not 0x0014)",
+                        size
+                    );
+                }
+                let mut guid = [0u8; 16];
+                bytes.read_exact(&mut guid)?;
+                Ok(DevicePath::FirmwareFile { guid })
+            }
+            (4, 7) => {
+                // "PIWG Firmware Volume" aka "Firmware Volume" in UEFI PI spec
+                let size = bytes.read_u16::<LittleEndian>()?;
+                if size != 0x14 {
+                    bail!("Firmware Volume size is wrong (was {:#04x}, not 0x0014)", size);
+                }
+                let mut guid = [0u8; 16];
+                bytes.read_exact(&mut guid)?;
+                Ok(DevicePath::FirmwareVolume { guid })
+            }
+            (ty, subtype) => {
+                bail!(
+                    "Device path type/subtype unsupported: ({:#02x}/{:#02x})",
+                    ty,
+                    subtype
+                );
+            }
+        }
+    }
+}
+
+impl EfiLoadOption {
+    // parsing here brought to you by rereading
+    // * https://uefi.org/specs/UEFI/2.10/10_Protocols_Device_Path_Protocol.html
+    // * https://uefi.org/specs/UEFI/2.10/03_Boot_Manager.html
+    pub(crate) fn parse_from(
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<EfiLoadOption, Error> {
+        let _attributes = bytes.read_u32::<LittleEndian>()?;
+        let file_path_list_length = bytes.read_u16::<LittleEndian>()?;
+
+        // The `Description` field is a null-terminated string of char16.
+        let mut description_chars: Vec<u16> = Vec::new();
+
+        loop {
+            let c = bytes.read_u16::<LittleEndian>()?;
+            description_chars.push(c);
+            if c == 0 {
+                break;
+            }
```

Comment:
Huh, this appears to include the null terminator in the Rust string we turn this into, is that actually the desired behavior?

Reply:
ah, no, it's not really. i'd collected bytes and decided to turn them into `String` much later, this was an oversight along the way. good catch

Reply:
(fixed in https://github.com/oxidecomputer/propolis/pull/756/commits/ed455ea695be0b6b03a7e39212ae6ec5b84b9fdc#diff-2a37b514b91da3540f6c06fa43cd50e589c319a328461ab05a7ea999a7bc162bL216)